File: Week 6 Python/Solution/150395760/sentimental-readability/readability.py
from cs50 import get_string

# 1. Obtener el texto
text = get_string("Text: ")

# 2. Inicializar contadores
letters = 0
sentences = 0
words = 0

# 3. Analizar el texto
# Para contar letras, iteramos sobre cada caracter
for char in text:
    if char.isalpha():
        letters += 1

# .split() crea una lista ['Hola', 'mundo'], len() nos dice cuántas hay
words = len(text.split())

# Para contar oraciones, buscamos los signos de puntuacion
sentences = text.count('.') + text.count('!') + text.count('?')

# 4. Calcular el indice Coleman-Liau
# L = letras por 100 palabras
L = (letters / words) * 100
# S = oraciones por 100 palabras
S = (sentences / words) * 100

index = 0.0588 * L - 0.296 * S - 15.8

# 5. Redondear al entero mas cercano
grade = round(index)

# 6. Imprimir el resultado segun las reglas
if grade < 1:
    print("Before Grade 1")
elif grade >= 16:
    print("Grade 16+")
else:
    print(f"Grade {grade}")

File: Week 2 Arrays/readability.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <math.h>

int count_letters(const char *text)
{
    int count = 0;
    for (size_t i = 0; text[i] != '\0'; ++i)
    {
        if (isalpha((unsigned char)text[i]))
            count++;
    }
    return count;
}

int count_words(const char *text)
{
    int count = 0;
    size_t len = strlen(text);
    if (len == 0) return 0;
    for (size_t i = 0; text[i] != '\0'; ++i)
    {
        if (text[i] == ' ')
            count++;
    }
    return count + 1;
}

int count_sentences(const char *text)
{
    int count = 0;
    for (size_t i = 0; text[i] != '\0'; ++i)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
            count++;
    }
    return count;
}

int main(void)
{
    char text[4096];

    printf("Texto: ");
    if (!fgets(text, sizeof(text), stdin)) return 1;
    text[strcspn(text, "\r\n")] = '\0';

    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    if (words == 0)
    {
        printf("Before Grade 1\n");
        return 0;
    }

    double L = (double)letters / words * 100.0;
    double S = (double)sentences / words * 100.0;

    double index = 0.0588 * L - 0.296 * S - 15.8;
    int grade = (int)round(index);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", grade);
    }

    return 0;
}
